In [1]:
import numpy as np
import os
import librosa
from tqdm.notebook import tqdm
import csv
import json

In [3]:
Report = {}

In [16]:
data_processed = '/Users/joemorford/Desktop/Sound stuff/from_Ines_020420/data_processed/'

raw_annotations_path ='/Users/joemorford/Desktop/Sound stuff/from_Ines_020420/Annotations'

labels_dict = {1:'male_in', 2:'female_in', 3:'chick', 4:'male_bout', 5:'female_bout', 6:'flapping', 7: 'male_grunt', 8:'female_grunt', 9:'unk_grunt' }

In [19]:
list_wav_files = os.listdir(data_processed)
n_files = len(list_wav_files)-1
print('n_files', str(n_files))
Report['n_files'] = n_files

n_files 52


In [11]:
# n_burrows:

burrows_ids = []
for filename in list_wav_files:
    burrow_id = filename.split('_')[0]
    burrows_ids.append(burrow_id)
    
burrows_ids = set(burrows_ids)
n_burrows = len(burrows_ids)
print('n_burrows', n_burrows)

n_burrows 12


In [24]:
dataset = np.load(data_processed+'dataset_25_03.npy', allow_pickle=True)

In [25]:
print('dataset shape: ', dataset.shape)
n_segments = dataset.shape[0]
print('n_segments', n_segments)
Report['n_segments']= n_segments

dataset shape:  (38049, 3)
n_segments 38049


## summary:
number audio recordings: 52

number of 6sec segments: 38049 

number of burrows: 12


## Numbers per burrow


In [92]:
# n_segments per burrow: 
# n_negative per burrow: number of segments of that burrow that are don't have any class
# n_segm_per_class_per_burrow
# 1:'male_in', 2:'female_in', 3:'chick', 4:'male_bout', 5:'female_bout', 6:'flapping', 7: 'male_grunt', 8:'female_grunt', 9:'unk_grunt' 
burrows_dict = {}
for segm in dataset:
    burrow_id = segm[0].split('_')[0]
    label_matrix = segm[-1]
    if burrow_id in burrows_dict.keys():
        burrows_dict[burrow_id]['n_segments'] = burrows_dict[burrow_id]['n_segments']+1
        if sum(sum(label_matrix)) == 0:
            burrows_dict[burrow_id]['n_neg_segments'] = burrows_dict[burrow_id]['n_neg_segments']+1
        active_classes = np.where(np.sum(label_matrix,1) > 0 )
        #print(active_classes)
        #print(np.sum(label_matrix,1))
        for i in active_classes[0]:
            if i ==0:
                burrows_dict[burrow_id]['n_seg_male_in'] = burrows_dict[burrow_id]['n_seg_male_in'] + 1
            if i ==1:
                 burrows_dict[burrow_id]['n_seg_female_in'] = burrows_dict[burrow_id]['n_seg_female_in'] + 1
            if i ==2:
                 burrows_dict[burrow_id]['n_seg_chick'] = burrows_dict[burrow_id]['n_seg_chick'] + 1
            if i ==3:
                burrows_dict[burrow_id]['n_seg_male_bout'] = burrows_dict[burrow_id]['n_seg_male_bout'] + 1
            if i ==4:
                 burrows_dict[burrow_id]['n_seg_female_bout'] = burrows_dict[burrow_id]['n_seg_female_bout'] + 1
            if i ==5:
                 burrows_dict[burrow_id]['n_seg_flapping'] = burrows_dict[burrow_id]['n_seg_flapping'] + 1
            if i ==6:
                 burrows_dict[burrow_id]['n_seg_male_grunt'] = burrows_dict[burrow_id]['n_seg_male_grunt'] + 1
            if i ==7:
                 burrows_dict[burrow_id]['n_seg_female_grunt'] = burrows_dict[burrow_id]['n_seg_female_grunt'] + 1
            if i ==8:
                 burrows_dict[burrow_id]['n_seg_unk_grunt'] = burrows_dict[burrow_id]['n_seg_unk_grunt'] + 1 
    else:
        burrows_dict[burrow_id]= {'n_segments':1, 'n_neg_segments':0, \
                                  'n_seg_male_in':0, 'n_seg_female_in':0, \
                                  'n_seg_chick':0, 'n_seg_male_bout':0,\
                                  'n_seg_female_bout':0, 'n_seg_flapping':0,\
                                  'n_seg_male_grunt':0, 'n_seg_female_grunt':0, 'n_seg_unk_grunt':0 }
        if sum(sum(label_matrix)) == 0:
            burrows_dict[burrow_id]['n_neg_segments'] = burrows_dict[burrow_id]['n_neg_segments'] +1
        active_classes = np.where(np.sum(label_matrix,1) > 0 )
        #print(active_classes)
        #print(np.sum(label_matrix,1))
        for i in active_classes[0]:
            if i ==0:
                burrows_dict[burrow_id]['n_seg_male_in'] = burrows_dict[burrow_id]['n_seg_male_in'] + 1
            if i ==1:
                 burrows_dict[burrow_id]['n_seg_female_in'] = burrows_dict[burrow_id]['n_seg_female_in'] + 1
            if i ==2:
                 burrows_dict[burrow_id]['n_seg_chick'] = burrows_dict[burrow_id]['n_seg_chick'] +1
            if i ==3:
                burrows_dict[burrow_id]['n_seg_male_bout'] = burrows_dict[burrow_id]['n_seg_male_bout'] + 1
            if i ==4:
                 burrows_dict[burrow_id]['n_seg_female_bout'] = burrows_dict[burrow_id]['n_seg_female_bout'] + 1
            if i ==5:
                 burrows_dict[burrow_id]['n_seg_flapping'] = burrows_dict[burrow_id]['n_seg_flapping'] + 1
            if i ==6:
                 burrows_dict[burrow_id]['n_seg_male_grunt'] = burrows_dict[burrow_id]['n_seg_male_grunt'] + 1
            if i ==7:
                 burrows_dict[burrow_id]['n_seg_female_grunit'] = burrows_dict[burrow_id]['n_seg_female_grunt'] + 1
            if i ==8:
                 burrows_dict[burrow_id]['n_seg_unk_grunt'] = burrows_dict[burrow_id]['n_seg_unk_grunt'] + 1   
import json
print(json.dumps(burrows_dict, indent=1))

{
 "3": {
  "n_segments": 4394,
  "n_neg_segments": 3794,
  "n_seg_male_in": 50,
  "n_seg_female_in": 80,
  "n_seg_chick": 351,
  "n_seg_male_bout": 48,
  "n_seg_female_bout": 70,
  "n_seg_flapping": 7,
  "n_seg_male_grunt": 16,
  "n_seg_female_grunt": 27,
  "n_seg_unk_grunt": 241
 },
 "135": {
  "n_segments": 6591,
  "n_neg_segments": 5884,
  "n_seg_male_in": 103,
  "n_seg_female_in": 34,
  "n_seg_chick": 555,
  "n_seg_male_bout": 93,
  "n_seg_female_bout": 28,
  "n_seg_flapping": 45,
  "n_seg_male_grunt": 44,
  "n_seg_female_grunt": 41,
  "n_seg_unk_grunt": 16
 },
 "206": {
  "n_segments": 498,
  "n_neg_segments": 483,
  "n_seg_male_in": 2,
  "n_seg_female_in": 0,
  "n_seg_chick": 12,
  "n_seg_male_bout": 2,
  "n_seg_female_bout": 0,
  "n_seg_flapping": 1,
  "n_seg_male_grunt": 0,
  "n_seg_female_grunt": 0,
  "n_seg_unk_grunt": 0
 },
 "49": {
  "n_segments": 5693,
  "n_neg_segments": 5042,
  "n_seg_male_in": 64,
  "n_seg_female_in": 3,
  "n_seg_chick": 557,
  "n_seg_male_bout": 63,
 

In [71]:
"""
with open("analysis_per_burrow.json", 'w') as oputfile:
    json.dump(burrows_dict, oputfile) 
"""

## Numbers per class:

In [ ]:
## per class:
#n_segments_per_class
#n_frames_per_class
#total_number_events, mean_duration, variance, max_duration, min_duration, \

labels_duration_dict = {'male_in':[], 'female_in':[],\
               'chick':[], 'male_bout':[], \
               'female_bout':[],'flapping':[], \
               'male_grunt':[],'female_grunt':[],\
               'unk_grunt':[]}




# go back to raw annotations to get idea of whole events characteristics
for rawfilename in os.listdir(raw_annotations_path):
    with open(raw_annotations_path + '/' + rawfilename, 'r') as rawfile:
        reader = csv.reader(rawfile)
        for row in reader:
            try:
                label_value = int(row[1])
            except ValueError as e:
                print("annotation error, will make it as closest label!")
                label_value = int(np.floor(float(row[1])))
                print(row[1])
            label_str = labels_dict[int(label_value)]
            labels_duration_dict[label_str].append(float(row[2]))

In [42]:
labels_sumaary_dict = {'male_in':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0}, \
                       'female_in':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0},\
                       'chick':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0}, \
                       'male_bout':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0}, \
                       'female_bout':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0},\
                       'flapping':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0}, \
                       'male_grunt':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0},\
                       'female_grunt':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0},\
                       'unk_grunt':{'nb_events':0, 'mean_duration':0, 'duration_sd':0, 'max_duration':0, 'min_duration':0}}


for key in labels_duration_dict.keys():
    nb_events = len(labels_duration_dict[key])
    mean_duration = np.mean(labels_duration_dict[key])
    sd = np.std(labels_duration_dict[key])
    max_d = np.max(labels_duration_dict[key])
    min_d = np.min(labels_duration_dict[key])
    labels_sumaary_dict[key]['nb_events'] = nb_events
    labels_sumaary_dict[key]['mean_duration'] = mean_duration
    labels_sumaary_dict[key]['duration_sd'] = sd
    labels_sumaary_dict[key]['max_duration'] = max_d
    labels_sumaary_dict[key]['min_duration'] = min_d
    
print(json.dumps(labels_sumaary_dict, indent=1)) 


{
 "male_in": {
  "nb_events": 1436,
  "mean_duration": 0.4784877988697771,
  "duration_sd": 0.12285852088022219,
  "max_duration": 0.96,
  "min_duration": 2.0833e-05
 },
 "female_in": {
  "nb_events": 650,
  "mean_duration": 0.29336128204615386,
  "duration_sd": 0.10924073602081764,
  "max_duration": 1.013333333,
  "min_duration": 0.061333333
 },
 "chick": {
  "nb_events": 13676,
  "mean_duration": 0.17862212848676512,
  "duration_sd": 0.12229438160066465,
  "max_duration": 1.602,
  "min_duration": 0.0
 },
 "male_bout": {
  "nb_events": 400,
  "mean_duration": 6.85610260419,
  "duration_sd": 3.3604846045723837,
  "max_duration": 23.57666667,
  "min_duration": 2.0833e-05
 },
 "female_bout": {
  "nb_events": 191,
  "mean_duration": 5.054647796628273,
  "duration_sd": 2.4821833459500664,
  "max_duration": 12.986666667,
  "min_duration": 2.0833e-05
 },
 "flapping": {
  "nb_events": 149,
  "mean_duration": 2.2066984060402683,
  "duration_sd": 1.6300262077262861,
  "max_duration": 10.346666

In [82]:
"""
with open("analysis_per_class.json", 'w') as oputfile:
    json.dump(labels_sumaary_dict, oputfile) 
"""

## Overlaps:

In [145]:
burrow_overlap_dictionary = {}
overlap_lengths_dict = {'total_male_length':0, 'total_female_length':0, \
                        'total_chick_length':0, 'm_f_overlap_length':0, \
                        'm_chick_overlap_length':0, 'f_chick_overlap_length':0, \
                        'all3_overlap_length':0, 'mf_bout_overlap_length':0, \
                        'mf_inhale_overlap_length':0, 'mf_coord':0}

length_each = 6/259

for segm in dataset:
    burrow_id = segm[0].split('_')[0]
    label_matrix = segm[-1]
    if burrow_id not in burrow_overlap_dictionary.keys():
        burrow_overlap_dictionary[burrow_id] = overlap_lengths_dict.copy()
    
    chick_inds = np.where(label_matrix[2]==1)[0]
    chick_n = len(chick_inds)
    burrow_overlap_dictionary[burrow_id]['total_chick_length'] +=  chick_n*length_each
    
    male_vec = label_matrix[0] + label_matrix[3] + label_matrix[6]
    male_inds = np.where(male_vec>0)[0]
    male_n = len(male_inds)
    burrow_overlap_dictionary[burrow_id]['total_male_length'] +=  male_n*length_each
    
    female_vec = label_matrix[1] + label_matrix[4] + label_matrix[7]
    female_inds = np.where(female_vec>0)[0]
    female_n = len(female_inds)
    burrow_overlap_dictionary[burrow_id]['total_female_length'] +=  female_n*length_each
    
    m_f_inds = set(male_inds).intersection(set(female_inds))
    m_f_n = len(m_f_inds)
    burrow_overlap_dictionary[burrow_id]['m_f_overlap_length'] +=  m_f_n*length_each

    m_chick_inds = set(male_inds).intersection(set(chick_inds))
    m_chick_n = len(m_chick_inds)
    burrow_overlap_dictionary[burrow_id]['m_chick_overlap_length'] +=  m_chick_n*length_each
    
    f_chick_inds = set(female_inds).intersection(set(chick_inds))
    f_chick_n = len(f_chick_inds)
    burrow_overlap_dictionary[burrow_id]['f_chick_overlap_length'] +=  f_chick_n*length_each
    
    inds_all3 = set(m_chick_inds).intersection(set(female_inds))
    all3_n = len(inds_all3)
    burrow_overlap_dictionary[burrow_id]['all3_overlap_length'] +=  all3_n*length_each

    mf_bouts = label_matrix[3] + label_matrix[4]
    mf_bouts_inds = np.where(mf_bouts==2)[0]
    mf_bout_n = len(mf_bouts_inds)
    burrow_overlap_dictionary[burrow_id]['mf_bout_overlap_length'] +=  mf_bout_n*length_each

    mf_in = label_matrix[0] + label_matrix[1]
    mf_in_inds = np.where(mf_in==2)[0]
    mf_in_n = len(mf_in_inds)
    burrow_overlap_dictionary[burrow_id]['mf_inhale_overlap_length'] +=  mf_in_n*length_each

for burrow in burrow_overlap_dictionary:
    bout_overlap = burrow_overlap_dictionary[burrow]['mf_bout_overlap_length']
    inhale_overlap = burrow_overlap_dictionary[burrow]['mf_inhale_overlap_length']
    if inhale_overlap>0:
        burrow_overlap_dictionary[burrow]['mf_coord'] = inhale_overlap/bout_overlap
    
print(json.dumps(burrow_overlap_dictionary, indent=1)) 

{
 "3": {
  "total_male_length": 144.4169884169885,
  "total_female_length": 203.37451737451732,
  "total_chick_length": 339.08108108108047,
  "m_f_overlap_length": 26.687258687258687,
  "m_chick_overlap_length": 1.528957528957529,
  "f_chick_overlap_length": 2.1312741312741315,
  "all3_overlap_length": 0.23166023166023167,
  "mf_bout_overlap_length": 22.12355212355212,
  "mf_inhale_overlap_length": 0.9266409266409266,
  "mf_coord": 0.041884816753926704
 },
 "135": {
  "total_male_length": 290.38610038610034,
  "total_female_length": 94.16988416988417,
  "total_chick_length": 519.4980694980693,
  "m_f_overlap_length": 6.671814671814671,
  "m_chick_overlap_length": 11.768339768339768,
  "f_chick_overlap_length": 8.895752895752896,
  "all3_overlap_length": 0.2084942084942085,
  "mf_bout_overlap_length": 5.444015444015443,
  "mf_inhale_overlap_length": 0.7644787644787645,
  "mf_coord": 0.14042553191489363
 },
 "206": {
  "total_male_length": 8.687258687258687,
  "total_female_length": 0.0


## Adult vs chick dataset:

    this will be the same as the main dataset (above) but the label matrix were manipulated to have only 3 clasees: 
        -adult_grunts
        -adult_bouts
        -chicks
     

In [2]:
adult_chick_dataset_path = '/home/ines/Dropbox/QMUL/PHD/manx_shearwaters/data/data_processed/adult_vs_chick/dataset_adult_bouts_grunts_chicks.npy'

adult_chick_dataset= np.load(adult_chick_dataset_path, allow_pickle=True)

In [10]:
dataset_positivesonly = np.asarray([e for e in adult_chick_dataset if 1 in e[2]])
print('n_segments total :', adult_chick_dataset.shape[0])
print("n_segments containing events: ", dataset_positivesonly.shape[0])
print('n_segments wich are silent: (in this case silences = adult_chick_dataset - positives only) :', str(adult_chick_dataset.shape[0] - dataset_positivesonly.shape[0]))

n_segments total : 38049
n_segments containing events:  3385
n_segments wich are silent: (in this case silences = adult_chick_dataset - positives only) : 34664


In [ ]:
del adult_chick_dataset
del dataset_positivesonly